src/backend/distributed/operations/citus_tools.c

```diff
@@ -52,14 +57,68 @@ static void ExecuteCommandsAndStoreResults(StringInfo *nodeNameArray,
 										   bool *statusArray,
 										   StringInfo *resultStringArray,
 										   int commandCount);
-static bool ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort,
-										char *queryString, StringInfo queryResult);
+static bool ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort, const char *user,
+										const char *database, char *queryString,
+										StringInfo queryResultString,
+										bool forceNewConnection);
 static Tuplestorestate * CreateTupleStore(TupleDesc tupleDescriptor,
 										  StringInfo *nodeNameArray, int *nodePortArray,
 										  bool *statusArray,
 										  StringInfo *resultArray, int commandCount);
 
 
+/*
+ * check_connection_to_node sends a simple query from a worker node to another
+ * node, and returns success status.
+ */
+Datum
+check_connection_to_node(PG_FUNCTION_ARGS)
+{
+	char *nodeName = PG_GETARG_TEXT_TO_CSTRING(0);
+	int32 nodePort = PG_GETARG_INT32(1);
+	char *userName = PG_GETARG_TEXT_TO_CSTRING_OR_NULL(2);
+	char *databaseName = PG_GETARG_TEXT_TO_CSTRING_OR_NULL(3);
+
+	bool success = CheckConnectionToNode(nodeName, nodePort, userName, databaseName);
+	PG_RETURN_BOOL(success);
+}
+
+
+/*
+ * CheckConnectionToNode sends a simple query from a node to another and returns success status
+ */
+static bool
+CheckConnectionToNode(char *nodeName, int32 nodePort, const char *userName,
+					  const char *databaseName)
+{
+	StringInfo queryResultString = makeStringInfo();
+	bool forceNewConnection = false;
+	volatile bool success = false;
+
+	/* ExecuteRemoteQueryOrCommand may throw errors in case of connection failures. */
+	PG_TRY();
+	{
+		success = ExecuteRemoteQueryOrCommand(nodeName, nodePort, userName, databaseName,
+											  CONNECTIVITY_CHECK_QUERY, queryResultString,
+											  forceNewConnection);
+	}
+	PG_CATCH();
+	{
+		success = false;
```

Comment:
We should be extremely careful about resuming after a PG_CATCH and avoid it where possible, e.g. we could be masking an OOM that messed up session-level state here.

Is there a way to use one of the ExecuteOptional commands?

Reply:
I managed to get `ExecuteOptionalRemoteCommand` to work.

However, the behaviour changed slightly as we no longer can intercept cancellations.

I guess we are ok with this change.

```diff
 -- cancel the connection after connectivity check query is sent
SELECT citus.mitmproxy('conn.onQuery(query="^SELECT 1$").cancel(' || pg_backend_pid() || ')');
 mitmproxy
 ---------------------------------------------------------------------

(1 row)

SELECT * FROM check_connection_to_node('localhost', :worker_2_proxy_port);
-DEBUG:  check_connection_to_node was not successful for localhost:xxxxx
-DETAIL:
- success
----------------------------------------------------------------------
- f
-(1 row)
+ERROR:  canceling statement due to user request
```

Reply:
This is probably the more expected user experience anyway